### ISSUE TITLE
clarify the stance on minification/obfuscation and non-symbolic stack traces

### ISSUE THREAD

**Issue by mraleph** [dart-lang/language#1212]
Dart VM has two execution modes where we hide or mangle symbolic information associated with the program:

* Obfuscated AOT produces stack traces which look like normal Dart stack traces, but with original names replaced by meaningless combinations of letters.
* AOT with split debug info produces stack traces which look more like stack traces you'd normally get from a native crash (with PC offsets rather than symbolic information).

Currently in both of these modes VM emits a warning saying "Warning: This VM has been configured to produce stack traces that violate the Dart standard." - I think this warning is confusing and unnecessary. 

So I would like the language team to issue a ruling clarifying the stance on this, e.g. by including special wording in the specification (or library documentation), which explicitly permits these alterations by the execution environment.

/cc @leafpetersen @lrhn @eernstg @sstrickl @mkustermann

**Comment by mkustermann** [comment]
@mraleph Additionally, we might want to replace the warning printed with a link to a dart.dev / flutter.dev article explaining why the stack trace is not readable and how to make it readable.

**Comment by eernstg** [comment]
I think the VM should simply stop indicating that there is a violation of the Dart standard (assuming that this refers to the specification of the language): The language specification only gives 3 pieces of information about the contents of a stack trace, and they do not specify any properties of the stack trace object:

- `with a stack trace corresponding to the current system state`, in section [Errors and Warnings](https://github.com/dart-lang/language/blob/67a658c9260ed229c9a32d4725dddcf3493cec6e/specification/dartLangSpec.tex#L890), specifying what it means to throw an exception.
- `let $t$ be a stack trace corresponding to the current execution state`, in section [Throw](https://github.com/dart-lang/language/blob/67a658c9260ed229c9a32d4725dddcf3493cec6e/specification/dartLangSpec.tex#L9332).
- `stack trace corresponding to the current execution state at the assertion`, in section [Assert](https://github.com/dart-lang/language/blob/67a658c9260ed229c9a32d4725dddcf3493cec6e/specification/dartLangSpec.tex#L16044).

So there is no promise that we can do anything other than `toString()` on the stack trace object, and there is no detailed promise about what that string contains, we only require the stack trace object to 'correspond' to the execution state.

I believe that it is fair to say that a stack trace string listing minimized names as well as one that shows PC offsets do correspond to the execution state.

That said, it may still be appropriate and helpful to explain the situation briefly when these special stack traces are used, at least as part of the stack trace string itself.

**Comment by leafpetersen** [comment]
@eernstg Agreed.

**Comment by lrhn** [comment]
Agree. The specification *deliberately* does not specify what a stack trace object is, nor does the `StackTrace` class. It has a `toString` so that it can be *printed*, but what it prints is entirely implementation specific.
That allows us to hand out whatever JavaScript give us without needing to post-process it.

It might be valuable to cooperate with the maintainers of `package:stacktrace` to make any other stack trace we produce parsable by that package, because it's likely where users will turn when they want to process stack traces.

**Comment by rmacnak-google** [comment]
The standard does not specify the exact format of stack traces, but it does impose several constraints on what must be included and excluded: functions may not be omitted because of inlining, synthetic functions may not be included, positions must be precise (line number alone is inadequate). This prohibits both obfuscation and external symbolization.

**Comment by lrhn** [comment]
@rmacnak-google 
The specification says *nothing* about the content of stack traces. The entirety of what it says is that it creates "a stack trace corresponding to the current system state".

The `StackTrace` class [documentation](https://api.dart.dev/stable/2.9.3/dart-core/StackTrace/toString.html) says one thing that might hint in that direction:
> The string represents the full stack trace starting from the point where a throw occurred to the top of the current call sequence.

It does not say *how* it represents that. It does not say that the "stack trace" must contain all methods invoked (tail call elimination would mean that the *stack* does not, but representing that faithfully would be correct).
The only reason it says "full" is that the VM has had a way to produce *partial* stack traces, and we didn't want it to be those.

There are no requirements on what as stack trace's `toString` contains.
*Users* tend to want as much information as possible, and they can get confused by missing frames, and that's a reason to give it to them. 
I would be perfectly fine with stack traces containing lines like:
```
--- omitted frames ---
```
for tail call elimination (which should be doable by setting one bit in the stack frame when doing a tail call), and with frames missing from AoT compiled code due to inlining (for JS, the source map can still point to the correct code, but you might lose the surrounding frame).

**Comment by rmacnak-google** [comment]
False.

```
The active stack trace is an object whose toString() method produces a string
that is a record of exactly those function activations within the current isolate
that had not completed execution at the point where the current exception
(16.9) was thrown.
This implies that no synthetic function activations may be added to the trace,
nor may any source level activations be omitted. This means, for example, that
any inlining of functions done as an optimization must not be visible in the trace.
Similarly, any synthetic routines used by the implementation must not appear in the
trace.
Nothing is said about how any native function calls may be represented in the
trace.
Note that we say nothing about the identity of the stack trace, or what notion
of equality is defined for stack traces.
The term position should not be interpreted as a line number, but rather as
a precise position - the exact character index of the expression that raised the
exception.
```

**Comment by lrhn** [comment]
The text does not appear in the [most recent](https://github.com/dart-lang/language/blob/master/specification/dartLangSpec.tex) versions of the language specification.
It was remived in https://github.com/dart-lang/sdk/commit/843bb96813f9a171c7257fcd5e282ff1d323c53d (Nov 25, 2016).

So, true! :wink:

**Comment by rmacnak-google** [comment]
The quoted text is from the [most recent](http://www.ecma-international.org/publications/standards/Ecma-408.htm) version of the language, which remains the 4th edition. Dart is specified by ECMA, not dart-lang/sdk.

**Comment by munificent** [comment]
> The quoted text is from the most recent version of the language, which remains the 4th edition.

That specification also defines generalized tear-offs. Are you planning to add those back to the VM? When you do, will you also be removing support for the Dart 2.0 type system, generic methods, optional `new`, optional `const`, mixin declaration syntax, set literals, spread, collection if, collection for, logical operators in const expressions, and extension members?

If not, I wouldn't get too hung up the ECMA-408 4th edition specification of stack traces. We decided several years go to stop going through the ECMA process once it became clear that Dart was not going to become natively embedded in browsers. That means that "the Dart team" is indeed not working on a compliant implementation of ECMA-408 and has not been for some years.

**Comment by leafpetersen** [comment]
@rmacnak-google I'm guessing your concern is not primarily with the spec but with the end user experience?  My personal take is that there is a tradeoff here - users would probably be marginally better served by deterministic stack traces across all platforms, but users are also better served by faster and smaller code.  Happy to hear more perspectives though - are there other considerations, or would you weight these concerns differently?

**Comment by mraleph** [comment]
@rmacnak-google Ryan, there is an interesting story behind the piece of standard that you are quoting.

You might notice a small issue with the quote. It says 

> The term position should not be interpreted as a line number, but rather as
a precise position - the exact character index of the expression that raised the
exception.

However if you look at the surrounding text you discover that term position does not actually contain the word **position** at all. Why is that? 

If you look at the commit that @lrhn has found you will discover that at some point spec was considering to be very strict about the content of the stack trace but this was regarded as a potential source of in-effeciency and consequently commented out. (Rightfully so because implementing that in JS would be impossible).

```latex
% Sadly, the info below cannot be computed efficiently. It would need to be computed at the throw point, since at latte points it might be destroyed. Native code in calling frames executes relative to the stack pointer, which therefore needs to be reset as each frame is unwound.  This means that the	
% OS kernel can dispose of this stack memory - it is not reliably preserved. And such code must execute if only to test if the exception should be caught or sent onward.	

% For each such function activation, the active stack trace includes the name of the function, the bindings of all its formal parameters, local variables and \THIS{}, and the position at which the function was executing.	

 % Is this controversial? We were thinking of viewing the trace as a List<Invocation>,	
 % but that won't capture the receiver or the locals. More generally, we need a standard interface that describes these traces, so one can type the stack trace variable in the catch.	


 \commentary{The term position should not be interpreted as a line number, but rather as a precise position - the exact character index of the  expression that raised  the exception. }	An \ON{}-\CATCH{} clause of the form \code{\CATCH{} ($p_1$, $p_2$) $s$} is equivalent to an \ON{}-\CATCH{} clause \code{\ON{} \DYNAMIC{} \CATCH{} ($p_1$, $p_2$) $s$}.

 % A position can be represented via a Token. If we make that part of the core reflection facility, we can state this here.
```

Whoever was commenting out part of the standard did not comment out obsolete commentary about position so it stayed behind - commenting on a no longer existing piece of text.

I'd argue that what is left in the standard, _specifies the content of the stack trace but not its representation_. We are just left with the following sentence:

> toString() method produces a string that is a record of exactly those function activations within the current isolate
that had not completed execution at the point where the current exception (16.9) was thrown.

There is nothing here that actually specifies how this record is represented, it just says that it should faithfully reproduce the state of the execution stack, which both obfuscated and non-symbolic stack traces do, they are "a record of exactly those function activations...". It's just non-human readable record, but one which can be demangled back into a corresponding human readable record without compromising its faithfulness (and thus without any negative impact on user experience).

(Small footnote: if we read too much into the old standard wording we would have to declare that our causal stack traces are also violating it, because suspended async functions are not represented as activations on the stack, but yet present in the stack trace).
